internal/mongodoc/resource.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongodoc // import "gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+
+import (
+	"fmt"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charm.v6-unstable/resource"
+)
+
+// CheckCharmResource ensures that the given entity is okay
+// to associate with a revisioned resource.
+func CheckCharmResource(entity *Entity, res resource.Resource) error {
+	// TODO(ericsnow) Verify that the revisioned resources is in the DB.
+
+	if err := res.Validate(); err != nil {
+		return err
+	}
+	if res.Fingerprint.IsZero() {
+		return errgo.Newf("resources must have a fingerprint")
+	}
+
+	if entity.URL.Series == "bundle" {
+		return errgo.Newf("bundles do not have resources")
+	}
+	if !charmHasResource(entity.CharmMeta, res.Name) {
+		return errgo.Newf("charm does not have resource %q", res.Name)
+	}
+
+	return nil
+}
+
+func charmHasResource(meta *charm.Meta, resName string) bool {
+	for name := range meta.Resources {
+		if resName == name {
+			return true
+		}
+	}
+	return false
+}
+
+// NewResourceID generates the doc ID corresponding to the given info.
+func NewResourceID(curl *charm.URL, resName string, revision int) string {
+	// We ignore the series and revision because resources are specific
+	// to the charm rather than to any particular variation of it.
+	curl = curl.WithRevision(-1)
+	curl.Series = ""
+	return fmt.Sprintf("resource#%s#%s#%d", curl, resName, revision)
+}
+
+// Resource holds the in-database representation of a charm resource
+// at a particular revision.
+type Resource struct {
+	DocID    string     `bson:"_id"`
```

Comment:
As discussed online, how about using a dual key here?

    Id ResourceId
    CharmURL *charm.URL

We can index them so they're unique in combination. It's possible
there's some other technical reason it's not good, but worth
trying, I think.

It's possible that ResourceId needs to be a pointer, but
again, worth experimenting with.